config/99-germ-analysis.py

```diff
@@ -0,0 +1,205 @@
+#germ_pd = germ_pd[T_START:NMAX+T_START]
+#germ_chip= germ_chip[T_START:NMAX+T_START]
+#germ_chan = germ_chan[T_START:NMAX+T_START]
+
+
+# make dask arrays
+#germ_ts = da.from_array(germ_ts, chunks=100)
+#germ_td = da.from_array(germ_td, chunks=100)
+#germ_pd = da.from_array(germ_pd, chunks=100)
+#germ_chip = da.from_array(germ_chip, chunks=100)
+#germ_chan = da.from_array(germ_chan, chunks=100)
+
+RESOLUTION = 40e-9 # seconds per clock cycle
+
+def fix_time(ts):	
+    ''' Fix the time in clock cycles from the FPGA 
+        taking into account the wrap.
+        ts : time (in clock cyles)
+            NOTE : edits in place
+
+        globals:
+            JUMP : the max number of clock cycles measured before jump
+            THRESH :
+                set some minimum threshold to detect the jump
+                don't want to add a jump at every negative diff
+                since the time stamps don't arrive exactly in order
+                (I was told it's on the order of 12 clock cyles, much less
+                less than my aggressive threhsold of 2**26)
+    '''
+    # copy and ensure it's int
+    #ts = ts.copy().astype(int)
+
+    # the jump is currently 2**27 (27 bits for clock ticks)
+    JUMP = 2**27
+    THRESH = 2**26
+
+    # time differences
+    ts_diff = np.diff(ts)
+
+    # find the regions where it's negative
+    # this should just be a few (1, 2 dozen maybe?)
+    w, = np.where(ts_diff < -THRESH)
+    if len(w) > 10:
+        print("Warning, detected more than 10 jumps? (long measurement maybe?)")
+        print("Detected {} jumps".format(len(w)))
+    njumps = 0
+    for ind in w:
+        njumps += 1
+        ts[ind+1:] += JUMP
+
+    return njumps
+
+    
+# for for jump in time
+# fixes array in place
+njumps = fix_time(germ_ts)
+# now sort in increasing order (because they're not exactly increasing)
+sort_ind = np.argsort(germ_ts)
+# and sort
+germ_ts = germ_ts[sort_ind]
+germ_td = germ_td[sort_ind]
+germ_pd = germ_pd[sort_ind]
+germ_chip = germ_chip[sort_ind]
+germ_chan = germ_chan[sort_ind]
+
+# debugging
+#plt.figure(2);plt.clf();plt.plot(germ_ts)
+#plt.pause(.0001)
+#input("check and confirm")
+
+tot_time = (germ_ts[-1]-germ_ts[0])*RESOLUTION
+print("Data is {}s long".format(tot_time))
+
+
+# now filter by time in seconds
+#time_delta = .1 # .1 s
+#time_delta_clocks = int(time_delta/RESOLUTION)
+
+# set to zero
+germ_ts_delta = germ_ts-germ_ts[0]
+# integer division for bins
+#germ_ts_delta_clocks = germ_ts_delta / time_delta_clocks
+
+n_chans =32
+
+chip_index = germ_chip*n_chans + germ_chan
+germ_pd_index = germ_pd
+time_index = germ_ts_delta #ids
+time_index = time_index*RESOLUTION
+
+# correct for the energy
+#corr_mat = 
+
+MAX_CHIP = np.max(chip_index)
+MAX_TIME = np.max(time_index)
+MAX_PD = 4096-1
+
+# chunk into 100 times
+N_TIMES = 100
+
+N_ENERGY = 4096
+MAX_ENERGY = 70
+
+from skbeam.core.accumulators.histogram import Histogram
+
+hh = Histogram((MAX_CHIP+1, 0, MAX_CHIP+1),
```

Comment:
Central piece is the use of scikit-beam `Histogram`

Reply:
Did you investigate doing this with pandas as well?

Reply:
no I did not. the idea here is to eventually be able to fill results on the fly.
Would pandas do that? I only see `.hist` which creates a 1D histogram.

Reply:
you can do this with groupby (like how it is done in heatmap).  `toolz` has support for streaming groupby I think....